### ISSUE TITLE
Stack Overflow 

### ISSUE THREAD

**Issue by flemminglau** [DependencyTrack/dependency-track#3201]
### Current Behavior

For many of our larger sbom uploads we are suddenly since yesterday getting this error and the upload does not happen:

```
2023-11-15 10:52:31,554 INFO [VulnerableSoftwareIndexer] Starting reindex task. This may take some time.
2023-11-15 10:52:31,554 INFO [IndexManager] Deleting vulnerablesoftware index
2023-11-15 10:52:45,861 INFO [VulnerableSoftwareIndexer] Reindexing complete
2023-11-15 10:52:59,260 INFO [BomUploadProcessingTask] Processing CycloneDX BOM uploaded to project: 28726d4a-c75b-4aac-9bc9-8590a6945dad
2023-11-15 10:53:12,420 INFO [BomUploadProcessingTask] Identified 16 new components
2023-11-15 10:53:12,420 INFO [BomUploadProcessingTask] Processing CycloneDX dependency graph for project: 28726d4a-c75b-4aac-9bc9-8590a6945dad
2023-11-15 10:53:13,898 ERROR [LoggableUncaughtExceptionHandler] An unknown error occurred in an asynchronous event or notification thread
java.lang.StackOverflowError: null
        at org.datanucleus.state.StateManagerImpl.transitionReadField(StateManagerImpl.java:1094)
        at org.datanucleus.state.StateManagerImpl.isLoaded(StateManagerImpl.java:4112)
        at org.dependencytrack.model.Component.dnGetchildren(Component.java)
        at org.dependencytrack.model.Component.getChildren(Component.java:687)
        at org.dependencytrack.persistence.ComponentQueryManager.recursivelyDelete(ComponentQueryManager.java:434)
        at org.dependencytrack.persistence.ComponentQueryManager.recursivelyDelete(ComponentQueryManager.java:436)
        at org.dependencytrack.persistence.ComponentQueryManager.recursivelyDelete(ComponentQueryManager.java:436)
        at org.dependencytrack.persistence.ComponentQueryManager.recursivelyDelete(ComponentQueryManager.java:436)
        at org.dependencytrack.persistence.ComponentQueryManager.recursivelyDelete(ComponentQueryManager.java:436)
        at org.dependencytrack.persistence.ComponentQueryManager.recursivelyDelete(ComponentQueryManager.java:436)
        at org.dependencytrack.persistence.ComponentQueryManager.recursivelyDelete(ComponentQueryManager.java:436)
        at org.dependencytrack.persistence.ComponentQueryManager.recursivelyDelete(ComponentQueryManager.java:436)
        at org.dependencytrack.persistence.ComponentQueryManager.recursivelyDelete(ComponentQueryManager.java:436)
        at org.dependencytrack.persistence.ComponentQueryManager.recursivelyDelete(ComponentQueryManager.java:436)
        at org.dependencytrack.persistence.ComponentQueryManager.recursivelyDelete(ComponentQueryManager.java:436)
        at org.dependencytrack.persistence.ComponentQueryManager.recursivelyDelete(ComponentQueryManager.java:436)
        at org.dependencytrack.persistence.ComponentQueryManager.recursivelyDelete(ComponentQueryManager.java:436)
        at org.dependencytrack.persistence.ComponentQueryManager.recursivelyDelete(ComponentQueryManager.java:436)
        at org.dependencytrack.persistence.ComponentQueryManager.recursivelyDelete(ComponentQueryManager.java:436)
        at org.dependencytrack.persistence.ComponentQueryManager.recursivelyDelete(ComponentQueryManager.java:436)
        at org.dependencytrack.persistence.ComponentQueryManager.recursivelyDelete(ComponentQueryManager.java:436)
        at org.dependencytrack.persistence.ComponentQueryManager.recursivelyDelete(ComponentQueryManager.java:436)
        at org.dependencytrack.persistence.ComponentQueryManager.recursivelyDelete(ComponentQueryManager.java:436)
        at org.dependencytrack.persistence.ComponentQueryManager.recursivelyDelete(ComponentQueryManager.java:436)
        at org.dependencytrack.persistence.ComponentQueryManager.recursivelyDelete(ComponentQueryManager.java:436)
        at org.dependencytrack.persistence.ComponentQueryManager.recursivelyDelete(ComponentQueryManager.java:436)
        at org.dependencytrack.persistence.ComponentQueryManager.recursivelyDelete(ComponentQueryManager.java:436)
```

We tried removing the entire _data directory and restart but no improvement.

Our next step will be to restore the DB from a backup.

### Steps to Reproduce

This upload of the same (or a past version of the same) bom was sucessful. Even with v4.9.1

It is entirely possible that the SBOM loaded contains circular references.
However other projects can be loaded despite them also containing circular references.

 

### Expected Behavior

No stack overflows

### Dependency-Track Version

4.9.1

### Dependency-Track Distribution

Container Image

### Database Server

PostgreSQL

### Browser

Google Chrome

### Checklist

- [X] I have read and understand the [contributing guidelines](https://github.com/DependencyTrack/dependency-track/blob/master/CONTRIBUTING.md#filing-issues)
- [X] I have checked the [existing issues](https://github.com/DependencyTrack/dependency-track/issues) for whether this defect was already reported

**Comment by flemminglau** [comment]
The offending SBOMs are constructed by downloading some sboms from DT, merging them and loading them back into DT.

The merge is done using cyclonedx-cli merge.

**Comment by flemminglau** [comment]
We have restored the DB from before the incident and DT is back in working condition.

So (of course) the cause was some corruption of the DB data.

Potentially triggered by the _data disk running full but we cannot know. Maybe in reality the _data disk ran full due to the huge logs generated by the stack overflows.

**Comment by nscuro** [comment]
This will be fixed with #3357.
